File: Data Structures & Algorithms/maximum-subarray/submission-3.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n+1, vector<int>(2,0));
        dp[n-1][1]=dp[n-1][0]=nums[n-1];
        for(int i = n-2; i>=0; --i){
            dp[i][0]=max(nums[i], dp[i+1][0]+nums[i]);
            dp[i][1]=max(dp[i+1][1], dp[i][0]);
        }
        return dp[0][1];
    }
};

File: Data Structures & Algorithms/jump-game-ii/submission-3.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        left, right = 0, 0
        count = 0
        while right < len(nums)-1:
            newright = right
            for i in range(left, right+1):
                newright= max(newright, i+ nums[i])
            left = right+1
            right = newright
            count +=1
        return count